src/Symfony/Component/Messenger/Transport/KernelTerminate/KernelTerminateTransport.php

```diff
@@ -0,0 +1,49 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Transport\KernelTerminate;
+
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Transport\TransportInterface;
+
+class KernelTerminateTransport implements TransportInterface
+{
+    private $handlers = array();
+
+    /**
+     * {@inheritdoc}
+     */
+    public function receive(callable $handler): void
+    {
+        $this->handlers[] = $handler;
```

Comment:
Actually, you would rather store envelopes in the `send` method. 
`receive` should then call `$handler` for each registered envelopes.

(`receive` name can be a bit misleading, as it's actually in charge of retrieving messages to treat, usually in a loop and pass them to the `$handler`. Also, it's never called during a request by the framework (until now with this transport and the `kernel.terminate` listener), the [Worker does](https://github.com/symfony/symfony/blob/b399527ba1c5c20f8b0f4db2b6563304c31c38ac/src/Symfony/Component/Messenger/Worker.php#L42-L48)).

Reply:
Indeed, I couldn’t figure out where to put it, it was 50% chance and I got it wrong. 
Docs are not really clear about that, maybe I can contribute to them too.
Thanks for the clarification!

Reply:
Yep, agree with @ogizanagi.